File: VB/RegistrationForm.Lesson5/ViewModel/RegistrationViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO
Imports RegistrationForm.DataModel
Imports System
Imports System.ComponentModel
Imports System.Windows

Namespace RegistrationForm.ViewModel

    <POCOViewModel>
    Public Class RegistrationViewModel
        Implements IDataErrorInfo

        Public Shared Function Create() As RegistrationViewModel
            Return ViewModelSource.Create(Function() New RegistrationViewModel())
        End Function

        Protected Sub New()
            MinBirthday = New DateTime(Date.Now.Year - 100, 12, 31)
            MaxBirthday = New DateTime(Date.Now.Year - 1, 12, 31)
            If IsInDesignMode() Then
                InitializeInDesignMode()
            Else
                InitializeInRuntime()
            End If
        End Sub

        Private Sub InitializeInDesignMode()
            FirstName = "John"
            LastName = "Smith"
            Email = "John.Smith@JohnSmithMail.com"
            Password = "Password"
            ConfirmPassword = "Password"
            Birthday = New DateTime(1980, 1, 1)
            Gender = 1
        End Sub

        Private Sub InitializeInRuntime()
            Birthday = Nothing
            Gender = -1
        End Sub

        Public Overridable Property FirstName As String

        Public Overridable Property LastName As String

        Public Overridable Property Email As String

        Public Overridable Property Password As String

        Public Overridable Property ConfirmPassword As String

        Public Overridable Property Birthday As Date?

        Public Overridable Property MinBirthday As Date

        Public Overridable Property MaxBirthday As Date

        Public Overridable Property Gender As Integer

        Public Sub AddEmployee()
            Dim [error] As String = EnableValidationAndGetError()
            If Not Equals([error], Nothing) Then
                OnValidationFailed([error])
                Return
            End If

            AddNewEmployee(FirstName, LastName, Email, Password, Birthday.Value, Gender)
            OnValidationSucceeded()
        End Sub

        Protected Sub OnPasswordChanged()
            RaisePropertyChanged(Function(x) x.ConfirmPassword)
        End Sub

        Private Sub OnValidationSucceeded()
            GetService(Of IMessageBoxService).Show("Registration succeeded", "Registration Form", MessageBoxButton.OK)
        End Sub

        Private Sub OnValidationFailed(ByVal [error] As String)
            GetService(Of IMessageBoxService).Show("Registration failed. " & [error], "Registration Form", MessageBoxButton.OK)
        End Sub

        Private allowValidation As Boolean = False

        Private Function EnableValidationAndGetError() As String
            allowValidation = True
            Dim [error] As String = CType(Me, IDataErrorInfo).Error
            If Not String.IsNullOrEmpty([error]) Then
                RaisePropertiesChanged()
                Return [error]
            End If

            Return Nothing
        End Function

        Private ReadOnly Property [Error] As String Implements IDataErrorInfo.[Error]
            Get
                If Not allowValidation Then Return Nothing
                Dim [me] As IDataErrorInfo = CType(Me, IDataErrorInfo)
                Dim lError As String = [me](BindableBase.GetPropertyName(Function() FirstName)) & [me](BindableBase.GetPropertyName(Function() LastName)) & [me](BindableBase.GetPropertyName(Function() Email)) & [me](BindableBase.GetPropertyName(Function() Password)) & [me](BindableBase.GetPropertyName(Function() ConfirmPassword)) & [me](BindableBase.GetPropertyName(Function() Birthday)) & [me](BindableBase.GetPropertyName(Function() Gender))
                If Not String.IsNullOrEmpty(lError) Then Return "Please check inputted data."
                Return Nothing
            End Get
        End Property

        Private ReadOnly Property Item(ByVal columnName As String) As String Implements IDataErrorInfo.Item
            Get
                If Not allowValidation Then Return Nothing
                Dim firstNameProp As String = BindableBase.GetPropertyName(Function() FirstName)
                Dim lastNameProp As String = BindableBase.GetPropertyName(Function() LastName)
                Dim emailProp As String = BindableBase.GetPropertyName(Function() Email)
                Dim passwordProp As String = BindableBase.GetPropertyName(Function() Password)
                Dim confirmPasswordProp As String = BindableBase.GetPropertyName(Function() ConfirmPassword)
                Dim birthdayProp As String = BindableBase.GetPropertyName(Function() Birthday)
                Dim genderProp As String = BindableBase.GetPropertyName(Function() Gender)
                If Equals(columnName, firstNameProp) Then
                    If Equals(FirstName, Nothing) OrElse String.IsNullOrEmpty(FirstName) Then Return String.Format("You cannot leave the {0} field empty.", firstNameProp)
                ElseIf Equals(columnName, lastNameProp) Then
                    If Equals(LastName, Nothing) OrElse String.IsNullOrEmpty(LastName) Then Return String.Format("You cannot leave the {0} field empty.", lastNameProp)
                ElseIf Equals(columnName, emailProp) Then
                    If Equals(Email, Nothing) OrElse String.IsNullOrEmpty(Email) Then Return String.Format("You cannot leave the {0} field empty.", emailProp)
                ElseIf Equals(columnName, passwordProp) Then
                    If Equals(Password, Nothing) OrElse String.IsNullOrEmpty(Password) Then Return String.Format("You cannot leave the {0} field empty.", passwordProp)
                ElseIf Equals(columnName, confirmPasswordProp) Then
                    If Not String.IsNullOrEmpty(Password) AndAlso Not Equals(Password, ConfirmPassword) Then Return "These passwords do not match. Please try again."
                ElseIf Equals(columnName, birthdayProp) Then
                    If Birthday Is Nothing OrElse String.IsNullOrEmpty(Birthday.ToString()) Then Return String.Format("You cannot leave the {0} field empty.", birthdayProp)
                ElseIf Equals(columnName, genderProp) Then
                    If Gender = -1 Then Return String.Format("You cannot leave the {0} field empty.", genderProp)
                End If

                Return Nothing
            End Get
        End Property
    End Class
End Namespace

File: CS/RegistrationForm.Lesson5/ViewModel/RegistrationViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.POCO;
using RegistrationForm.DataModel;
using System;
using System.ComponentModel;
using System.Windows;

namespace RegistrationForm.ViewModel {
    [POCOViewModel]
    public class RegistrationViewModel : IDataErrorInfo {
        public static RegistrationViewModel Create() {
            return ViewModelSource.Create(() => new RegistrationViewModel());
        }
        protected RegistrationViewModel() {
            MinBirthday = new DateTime(DateTime.Now.Year - 100, 12, 31);
            MaxBirthday = new DateTime(DateTime.Now.Year - 1, 12, 31);
            if(this.IsInDesignMode())
                InitializeInDesignMode();
            else InitializeInRuntime();
        }
        void InitializeInDesignMode() {
            FirstName = "John";
            LastName = "Smith";
            Email = "John.Smith@JohnSmithMail.com";
            Password = "Password";
            ConfirmPassword = "Password";
            Birthday = new DateTime(1980, 1, 1);
            Gender = 1;
        }
        void InitializeInRuntime() {
            Birthday = null;
            Gender = -1;
        }

        public virtual string FirstName { get; set; }
        public virtual string LastName { get; set; }
        public virtual string Email { get; set; }
        public virtual string Password { get; set; }
        public virtual string ConfirmPassword { get; set; }
        public virtual DateTime? Birthday { get; set; }
        public virtual DateTime MinBirthday { get; set; }
        public virtual DateTime MaxBirthday { get; set; }
        public virtual int Gender { get; set; }

        public void AddEmployee() {
            string error = EnableValidationAndGetError();
            if(error != null) {
                OnValidationFailed(error);
                return;
            }
            EmployeesModelHelper.AddNewEmployee(FirstName, LastName, Email, Password, Birthday.Value, Gender);
            OnValidationSucceeded();
        }
        protected void OnPasswordChanged() {
            this.RaisePropertyChanged(x => x.ConfirmPassword);
        }
        void OnValidationSucceeded() {
            this.GetService<IMessageBoxService>().Show("Registration succeeded", "Registration Form", MessageBoxButton.OK);
        }
        void OnValidationFailed(string error) {
            this.GetService<IMessageBoxService>().Show("Registration failed. " + error, "Registration Form", MessageBoxButton.OK);
        }

        bool allowValidation = false;
        string EnableValidationAndGetError() {
            allowValidation = true;
            string error = ((IDataErrorInfo)this).Error;
            if(!string.IsNullOrEmpty(error)) {
                this.RaisePropertiesChanged();
                return error;
            }
            return null;
        }
        string IDataErrorInfo.Error {
            get {
                if(!allowValidation) return null;
                IDataErrorInfo me = (IDataErrorInfo)this;
                string error =
                    me[BindableBase.GetPropertyName(() => FirstName)] +
                    me[BindableBase.GetPropertyName(() => LastName)] +
                    me[BindableBase.GetPropertyName(() => Email)] +
                    me[BindableBase.GetPropertyName(() => Password)] +
                    me[BindableBase.GetPropertyName(() => ConfirmPassword)] +
                    me[BindableBase.GetPropertyName(() => Birthday)] +
                    me[BindableBase.GetPropertyName(() => Gender)];
                if(!string.IsNullOrEmpty(error))
                    return "Please check inputted data.";
                return null;
            }
        }
        string IDataErrorInfo.this[string columnName] {
            get {
                if(!allowValidation) return null;
                string firstNameProp = BindableBase.GetPropertyName(() => FirstName);
                string lastNameProp = BindableBase.GetPropertyName(() => LastName);
                string emailProp = BindableBase.GetPropertyName(() => Email);
                string passwordProp = BindableBase.GetPropertyName(() => Password);
                string confirmPasswordProp = BindableBase.GetPropertyName(() => ConfirmPassword);
                string birthdayProp = BindableBase.GetPropertyName(() => Birthday);
                string genderProp = BindableBase.GetPropertyName(() => Gender);
                if (columnName == firstNameProp) {
                    if (FirstName == null || string.IsNullOrEmpty(FirstName))
                        return string.Format("You cannot leave the {0} field empty.", firstNameProp);
                } else if (columnName == lastNameProp) {
                    if (LastName == null || string.IsNullOrEmpty(LastName))
                        return string.Format("You cannot leave the {0} field empty.", lastNameProp);
                } else if (columnName == emailProp) {
                    if (Email == null || string.IsNullOrEmpty(Email))
                        return string.Format("You cannot leave the {0} field empty.", emailProp);
                } else if (columnName == passwordProp) {
                    if (Password == null || string.IsNullOrEmpty(Password))
                        return string.Format("You cannot leave the {0} field empty.", passwordProp);
                } else if (columnName == confirmPasswordProp) {
                    if (!string.IsNullOrEmpty(Password) && Password != ConfirmPassword)
                        return "These passwords do not match. Please try again.";
                } else if (columnName == birthdayProp) {
                    if (Birthday == null || string.IsNullOrEmpty(Birthday.ToString()))
                        return string.Format("You cannot leave the {0} field empty.", birthdayProp);
                } else if (columnName == genderProp) {
                    if (Gender == -1)
                        return string.Format("You cannot leave the {0} field empty.", genderProp);
                }
                return null;
            }
        }
    }
}